app/src/main/java/com/sdp13epfl2021/projmag/database/UserProfileDatabase.kt

```diff
@@ -0,0 +1,110 @@
+
+        /**
+         *  The field containing favorites
+         */
+        const val FAVORITES_FIELD = "profile"
+    }
+
+    /**
+     * Return the current logged user or null if none
+     *
+     * @return current logged user or null
+     */
+    private fun getUser(): FirebaseUser? = auth.currentUser
+
+
+    /**
+     * Return the document associated to the user
+     *
+     * @return document associated to the user
+     */
+    private fun getUserDoc(): DocumentReference? =
+        getUser()?.let { user ->
+            firestore
+                .collection(UserDataFirebase.ROOT)
+                .document(user.uid)
+        }
+
+    public fun uploadProfile(profile : ImmutableProfile) {
+        val profile = hashMapOf(
+            "firstName" to profile.firstName,
+            "lastName" to profile.lastName,
+            "age" to profile.age,
+            "gender" to profile.gender.toString(),
+            "phoneNumber" to profile.phoneNumber,
+            "role" to profile.role.toString(),
+            "sciper" to profile.sciper
+        )
+        firestore.collection(ROOT).document(getUser()?.uid!!)
+            .set(profile)
+            .addOnSuccessListener {
+                Log.d(TAG, "DocumentSnapshot added with ID:")
+            }
+            .addOnFailureListener {
+            }
+    }
+
+    public fun getProfile(onSuccess: (profile : ImmutableProfile?) -> Unit) {
+        val docRef =  firestore.collection(ROOT).document(getUser()?.uid!!)
+        docRef.get()
+            .addOnSuccessListener { document ->
+                if (document != null) {
+                    Log.d(TAG, "DocumentSnapshot data: ${document.data}")
+                    val firstName = document.data?.get("firstName")
+                    val lastName = document.data?.get("lastName")
+                    val ageLong = document.data?.get("age")
+                    val genderString = document.data?.get("gender")
+                    val gender = if(genderString?.equals(Gender.MALE.toString())!!) Gender.MALE else if(genderString?.equals(Gender.FEMALE.toString())!!)  Gender.FEMALE else Gender.OTHER
+                    val phoneNumber = document.data?.get("phoneNumber")
+                    val roleString = document.data?.get("role")
+                    val sciperLong = document.data?.get("sciper")
+                    val role = if (roleString == "Teacher") Role.TEACHER else if (roleString == "Student") Role.STUDENT else Role.OTHER
+                    val age = Integer.valueOf(ageLong.toString())
+                    var sciper : Int = 0
+                    try {
+                        sciper = Integer.valueOf(sciperLong.toString())
+                    } catch(exep : NumberFormatException) {
+                        sciper = 123456
+                    }
+
+
+                    val resProfile = ImmutableProfile.build(lastName as String, firstName as String, age, gender, sciper, phoneNumber as String, role)
```

Comment:
Unsafe cast could throw an exception, a try catch could avoid crashing the app if one field coulnd't be cast.
Or checking variable before, for example
```
if (firstName is String && ...) {
    val resProfile = ...
} else {
    ...
}
```